main/test/flix/Test.Dec.Class.flix

```diff
@@ -226,6 +226,24 @@ namespace Test/Dec/Class {
         instance D[F[true]]
     }
 
+    namespace Test19 {
+        pub enum M[_: Bool, a: Type] {
+          case M(a)
+        }
+
+        instance Eq[M[true, a]] with Eq[a] {
+            pub def eq(x: M[true, a], y: M[true, a]): Bool = match (x, y) {
+                case (M(x1), M(y1)) => x1 == y1
+            }
+        }
+
+        @test
+        def testCallComplexBoolSig01(): Bool = M(10): M[true, Int] == M(10): M[true, Int]
```

Comment:
Without the type annotation it becomes ambiguous right?

Reply:
Yes, because the instance `M[true, a]` has to generalize to the use `M[t, Int]`, which fails here.

Reply:
This becomes more of a problem with datalog constraints:

```scala
    @test
    def testStratification10(): Unit =
        // `_a` and `_b` will not be in the same stratification since they do not include
        // the `Diff_10` of the other.
        let _a = #{
            A10(x) :- B10(x), DiffA10(_).
        };
        let _b = #{
            B10(x) :- not A10(x), DiffB10(x).
        };
        ()
```

Here we have an error because the type of `x` is a variable and we can't confirm that it is `Boxable`.

Thoughts?

Reply:
I may have lost track. Whats the relation between this ticket and boxable?

Reply:
Suppose we want to check that type `T` has an instance of class `C`. Previously, we looked at all the instances `C[U]` and found the instance where `T unifiesWith U`. When `T` is a variable bound by a definition, there may be more than one such instance, but the selection of the right instance would be handled in `Monomorph`.

Now with complex instances, we need to be a little more specific and check that `U <= T` (i.e. `T` can be generalized to `U`). In adding this restriction I revealed what could be considered a bug in the current implementation. When `T` is a variable that is _not_ bound by a definition (as in the type of `x` in the constraints in `testStratification10` above), constraints may be inferred for `T` (in this case `Boxable[T]`) but these constraints are not in the class environment, since there is no reference to `T` in the signature of the function.

Previously this managed to sneak by, because `Boxable[a]` for some free variable `a` would unify with `Boxable[Unit]`, for example; the erroneous assumption was that multiple matches could only happen with overlapping or complex instances, which would be caught later. With the new `<=` requirement, we're catching this now.

Usually, this could be addressed by requiring the user to add a type ascription, such as `Nil: List[Int32]`. But with constraints this seems fairly unwieldy. I'll note further that this issue is not limited only to test cases: the [`half-siblings` koan](https://github.com/flix/flix/blob/master/examples/koans/half-siblings.flix) is also affected.

tl;dr: If we have a free type variable, no constraints from the class environment can be applied to it, so if constraints are inferred for it, we get an error. The way around this is type ascription, but it doesn't seem reasonable to require the user to do that, especially since it's not always obvious where such ascriptions would be needed.

Reply:
I may need "explain like I am 5".

This sounds like the classic `Nil == Nil` problem? Is it?

If so, the usually solution is to ban it and require type signatures. 

How many Datalog programs are actually affected by this?

In half-siblings where is the unbound Boxable coming from? Ok. I think its from the second argument to Parent.

Reply:
@JonathanStarup I am looking at the half-sibling program. Isnt that program a bit de-generate (or plain wrong if you will)? Where are the parent facts supposed to come from?

Reply:
It is the `Nil == Nil` issue. But more annoying because we can't just add ascriptions to constraint sets (unless I'm mistaken). A bunch of test cases and the half-siblings koans are the only programs I have discovered so far that are affected.